src/backend/mpi/swdsm.cpp

```diff
@@ -863,17 +920,37 @@ void argo_initialize(std::size_t argo_size, std::size_t cache_size){
 		vm::map_memory(tmpcache, offsets_tbl_size_bytes, current_offset, PROT_READ|PROT_WRITE);
 	}
 
-	sem_init(&ibsem, 0, 1);
+	mpi_windows = env::mpi_windows();
 
-	globalDataWindow = static_cast<MPI_Win*>(malloc(numtasks*sizeof(MPI_Win)));
+	// Create one data_window per page chunk
+	// TODO: Do we need the double dimensions or can each window be reused for another node?
+	data_windows.resize(mpi_windows, std::vector<MPI_Win>(numtasks));
+	for(std::size_t i = 0; i < mpi_windows; i++){
+		for(int j = 0; j < numtasks; j++){
+			MPI_Win_create(globalData, size_of_chunk, 1,
+						   MPI_INFO_NULL, MPI_COMM_WORLD, &data_windows[i][j]);
+		}
+	}
+	// Locks to protect the globalData windows from concurrent local access
+	mpi_lock_data = new mpi_lock*[mpi_windows];
+	for(std::size_t i = 0; i < mpi_windows; i++){
+		mpi_lock_data[i] = new mpi_lock[numtasks];
+	}
 
-	for(argo::node_id_t i = 0; i < numtasks; i++){
- 		MPI_Win_create(globalData, size_of_chunk*sizeof(argo_byte), 1,
-									 MPI_INFO_NULL, MPI_COMM_WORLD, &globalDataWindow[i]);
+	// Create one sharer_window per page chunk
+	sharer_windows.resize(mpi_windows, std::vector<MPI_Win>(numtasks));
+	for(std::size_t i = 0; i < mpi_windows; i++){
+		for(int j = 0; j < numtasks; j++){
+			MPI_Win_create(globalSharers, gwritersize, sizeof(unsigned long),
```

Comment:
Use some more appropriate type than `unsigned long`.

Reply:
This is now `sizeof(std::uint64_t)` as it should be.